tests/Game.class.test.js

```diff
@@ -0,0 +1,916 @@
+        [0, 0, 0, 0],
+      ]);
+    });
+
+    it('should have a score of 0 by default', () => {
+      const game2048 = new Game();
+
+      expect(game2048.getScore()).toBe(0);
+    });
+
+    it('should have a status of "idle" by default', () => {
+      const game2048 = new Game();
+
+      expect(game2048.getStatus()).toBe('idle');
+    });
+
+    it('should have a state of a custom board if the initial state is provided', () => {
+      const initialState = [
+        [2, 0, 0, 0],
+        [0, 4, 0, 0],
+        [0, 0, 8, 0],
+        [0, 0, 0, 16],
+      ];
+
+      const game2048 = new Game(cloneState(initialState));
+      const state = game2048.getState();
+
+      expect(state).toEqual(initialState);
+    });
+  });
+
+  describe('after start', () => {
+    it('should have a score of 0', () => {
+      const game2048 = new Game();
+
+      game2048.start();
+
+      expect(game2048.getScore()).toBe(0);
+    });
+
+    it('should have a status of "playing"', () => {
+      const game2048 = new Game();
+
+      game2048.start();
+
+      expect(game2048.getStatus()).toBe('playing');
+    });
+
+    it('should have a state with two cells with values 2 or 4', () => {
+      const game2048 = new Game();
+
+      game2048.start();
+
+      const flatState = game2048.getState().flat();
+      const nonEmptyCells = flatState.filter((cell) => cell !== 0);
+
+      expect(nonEmptyCells).toHaveLength(2);
+
+      expect(
+        nonEmptyCells.every((cell) => [2, 4].includes(cell)),
+      ).toBe(true);
+    });
+
+    it('should generate the board with 4s less often than with 2s', () => {
+      const game2048 = new Game();
+
+      let statesCombined = [];
+      let twos = 0;
+      let fours = 0;
+
+      for (let i = 0; i < 100; i++) {
+        game2048.start();
+
+        const flatState = game2048.getState().flat();
+
+        statesCombined = [
+          ...statesCombined,
+          ...flatState,
+        ];
+      }
+
+      statesCombined.forEach((cell) => {
+        if (cell === 2) {
+          twos++;
+        } else if (cell === 4) {
+          fours++;
+        }
+      });
+
+      expect(fours < twos).toBe(true);
+    });
+
+    it('should generate the board with cells in random positions', () => {
+      const game2048 = new Game();
+
+      const cellPositions = new Map();
+
+      for (let i = 0; i < 100; i++) {
+        game2048.start();
+
+        const state = game2048.getState();
+
+        state.forEach((row, rowIndex) => {
+          row.forEach((cell, cellIndex) => {
+            if (cell !== 0) {
+              const key = `${rowIndex}${cellIndex}`;
+
+              if (cellPositions.has(key)) {
+                cellPositions.set(key, cellPositions.get(key) + 1);
+              } else {
+                cellPositions.set(key, 1);
+              }
+            }
+          });
+        });
+      }
+
+      expect(cellPositions.size >= 8).toBe(true);
+      expect(cellPositions.size <= 16).toBe(true);
```

Comment:
Random test can fail randomly so in some rear cases it will fail without a reason

Reply:
I have increased the number of iterations for the test. However, to be honest, I think it's unnecessary. The test always returns 16 items with a good distribution in my case. Thus, I will set the `cellPositions.size` to be equal to 12 or greater since students should write their code correctly. For a properly implemented random position logic, this test will always be true.